php_phongo.c

```diff
@@ -1461,6 +1461,45 @@ static const char* php_phongo_bson_type_to_string(bson_type_t type) /* {{{ */
 		bson_iter_key(&(iter)),                        \
 		php_phongo_bson_type_to_string(bson_iter_type(&(iter))))
 
+static bool php_phongo_uri_finalize_auth(mongoc_uri_t* uri TSRMLS_DC) /* {{{ */
+{
+	const char *source = NULL;
+
+	source = mongoc_uri_get_auth_source(uri);
+
+	/* authSource with GSSAPI or X509 should always be external */
+	if (mongoc_uri_get_auth_mechanism(uri)) {
+		if (!strcasecmp(mongoc_uri_get_auth_mechanism(uri), "GSSAPI") ||
+			!strcasecmp(mongoc_uri_get_auth_mechanism(uri), "MONGODB-X509")) {
+			if (source) {
+				if (strcasecmp(source, "$external")) {
+					phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Failed to parse URI options: GSSAPI and X509 require \"$external\" authSource.");
+					return false;
+				}
+			}
+		}
+
+		/* MONGODB-X509 is the only mechanism that doesn't require username */
+		if (strcasecmp(mongoc_uri_get_auth_mechanism(uri), "MONGODB-X509") != 0) {
+			if (!mongoc_uri_get_username(uri) ||
+				strcmp(mongoc_uri_get_username(uri), "") == 0) {
+				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Failed to parse URI options: '%s' authentication mechanism requires username.", mongoc_uri_get_auth_mechanism(uri));
+				return false;
+			}
+		}
+
+		/* MONGODB-X509 errors if a password is supplied. */
+		if (strcasecmp(mongoc_uri_get_auth_mechanism(uri), "MONGODB-X509") == 0) {
+			if (mongoc_uri_get_password(uri)) {
+				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Failed to parse URI options: '%s' authentication mechanism does not accept a password.", mongoc_uri_get_auth_mechanism(uri));
```

Comment:
At this point we know the auth mechanism is `MONGODB-X509` so I think you can just write that directly into the error message. The only benefit I can see to using `mongoc_uri_get_auth_mechanism(uri)` is that it preserves the original case, which I don't think is too important.

I would suggest we use:

```
"Failed to parse URI options: X509 authentication mechanism does not accept a password."
```

Referring to "X509" instead of "MONGODB-X509" is also consistent with the first exception message in this function.